File: algorithm/003-longest_substring_without_repeating_characters/csharp/solution.cs
using System;

public class Solution {
    public int lengthOfLongestSubstring(string s) {
        int[] map = new int[127];
        int longest = 0;
        int i = 0; // 滑窗左边界
        int j = 0; // 滑窗右边界

        for ( ; j < s.Length; j++) {
            char c = s[j];
            if (map[c] > 0) {
                longest = Math.Max(longest, j - i);
                while (i < map[c]) {
                    map[s[i]] = 0;
                    i++;
                }
            }

            map[c] = j + 1;
        }

        return Math.Max(longest, j - i);
    }
}

File: algorithm/003-longest_substring_without_repeating_characters/kotlin/solution.kt
class Solution {
  fun lengthOfLongestSubstring(s: String): Int {
        val map = Array<Int>(127, {0})
        var longest = 0
        var i = 0 // 重复起始字符索引
        var j = 0

        for (c in s) {
            if (map[c.toInt()] > 0) {
                longest = Math.max(longest, j - i)
                while (i < map[c.toInt()]) {
                    map[s[i].toInt()] = 0
                    i++
                }
            }

            map[c.toInt()] = j + 1
            j++
        }

        return Math.max(longest, j - i)
    }
}